Auto gen Bash completion file

ArangoGutierrez opened the pull request:
**Description of the Pull Request (PR):**


**This fixes or addresses the following GitHub issues:**

- Ref: #
 - need a tab completion file #1890


**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [ ] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [x] I have tested this PR locally with a `make test`
- [x] This PR is against the project's master branch
- [x] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [x] This PR is ready for review and/or merge


Attn: @singularityware-admin

ArangoGutierrez committed:
Add etc/bash_completion.d to .gitignore file

Avoid commiting the auto gen bash_completion file

ArangoGutierrez committed:
Auto gen a bash_completion file

Use go gen to generate a bash_completion file for singularity CLI
under src/docs/bash_completion.d

ArangoGutierrez committed:
Add bash_completion file to the Makefile.stub

set rules on MakeFile to create and install the bash_completion file
for the singularity CLI.

ArangoGutierrez committed:
set LICENSE header on bash_completion.go file

jmstover commented on mlocal/frags/Makefile.stub:
```diff
@@ -71,6 +74,18 @@ $(go_OBJ): $(BUILDDIR)/vendors-done $(libruntime) $(BUILDDIR)/config.h
+$(bash_completion_INSTALL): $(bash_completion)
+	@echo " INSTALL" $@
+	$(V)install -d $(@D)
+	$(V)install -m 0755 $(bash_completion) $(bash_completion_INSTALL)
```
`install -m 0644 [...]`

Completion files are sourced, not executed.

ArangoGutierrez committed:
Fix merge conflicts against upstream/master

jmstover commented on mlocal/frags/Makefile.stub:
```diff
@@ -29,18 +29,21 @@ go_BIN := $(singularity)
 go_INSTALL := $(singularity_INSTALL)
 
+bash_completion :=  $(SOURCEDIR)/etc/bash_completion.d/singularity
+bash_completion_INSTALL := /etc/bash_completion.d/singularity
```
Should this be installed at the system `/etc/bash_completion.d` or at `$(PREFIX)/etc/bash_completion.d` ??  @yhcote

yhcote replied:
Well, I don't think bash will look into /usr/local, first impression is that bash expects its config from a tree off /etc/bash_completion.d/, but I might be wrong.

ArangoGutierrez replied:
is `$(PREFIX)/etc/bash_completion.d` sourced at boot?

ArangoGutierrez replied:
lol I was asking the same, for a matter of secs

jmstover replied:
I ask because even with autoconf, if you just define `--prefix` then _everything_ should be installed under that prefix. You'd need to do something like:

`./configure --prefix=/usr/local --sysconfdir=/etc` 

To get the behavior you have going here.

ArangoGutierrez replied:
from the documentation:
 - https://debian-administration.org/article/316/An_introduction_to_bash_completion_part_1

jmstover replied:
From my system `$HOME/.bash_profile`

```
    . /etc/bash_completion.d/git-completion.bash;
  [ ... ]
    . /usr/local/singularity/2.6.0/etc/bash_completion.d/singularity
```

So uhh... I'm sourcing each specific file I want. I do have a global profile script that sets up the generic completion helpers... but beyond that, none of the completion files get sourced automatically. I have to manually source them if I want them.

jmstover replied:
My expected behavior would be to have this install to:  $(PREFIX)/etc/bash_completion.d  ..... everything falling under $(PREFIX) since there currently isn't a way to specify bindir, sysconfdir, etc...

ArangoGutierrez replied:
@yhcote if you agree, I'll commit the requested change

ArangoGutierrez replied:
**NOTE** : if this change is done, the how to install the `bash_completion` file will have to be detailed on the install documentation, so users can setup sourcing the file on their `$HOME/.bash_profile` script

ArangoGutierrez committed:
remove bash_completion from CLEAN_FILES rule at Makefile.stub

ArangoGutierrez committed:
Completion files are sourced, not executed.

ArangoGutierrez committed:
jms was wrong about the CLEAN_FILES rule

yhcote commented on mlocal/frags/Makefile.stub:
```diff
@@ -31,18 +31,21 @@ go_BIN := $(singularity)
+INSTALLFILES := $(singularity_INSTALL) $(starter_INSTALL) $(starter_suid_INSTALL) $(sessiondir) $(config_INSTALL) $(dist_bin_SCRIPTS_INSTALL) $(capability_JSON) $(bash_completion_INSTALL)
 
-CLEANFILES += $(libruntime) $(starter) $(singularity) $(go_BIN) $(go_OBJ) 
+CLEANFILES += $(libruntime) $(starter) $(singularity) $(go_BIN) $(go_OBJ) $(bash_completion) 
```
Don't think you wanna remove your bash_completion source there.

yhcote commented on mlocal/frags/Makefile.stub:
```diff
@@ -73,6 +76,18 @@ $(go_OBJ): $(BUILDDIR)/vendors-done $(libruntime) $(BUILDDIR)/config.h
+# bash_completion file
+$(bash_completion): $(go_OBJ)
+	@rm -f $(bash_completion)
+	@mkdir -p $(SOURCEDIR)/etc/bash_completion.d 
```
If it's a generated file, it should end up in BUILDDIR not SOURCEDIR, that's why I commented earlier on the addition of source file to CLEANFILES.

yhcote commented on mlocal/frags/Makefile.stub:
```diff
@@ -31,18 +31,21 @@ go_BIN := $(singularity)
 go_INSTALL := $(singularity_INSTALL)
 
+bash_completion :=  $(SOURCEDIR)/etc/bash_completion.d/singularity
+bash_completion_INSTALL := /etc/bash_completion.d/singularity
```
As per the discussion, you still need to change bash_completion_INSTALL and prepend it with $(PREFIX).

yhcote reviewed (changes requested):
There's still a few things to change I think.

ArangoGutierrez committed:
change bash_completion install path

As the PR conversation states, change bash_completion_INSTALL rule
from `/etc/bash_completion.d/` to `$(PREFIX)/etc/bash_completion.d/`.

jmstover reviewed (dismissed):
Looks good here... 

```
$ . /usr/local/singularity/3.0.0-dev/etc/bash_completion.d/singularity 
$ /usr/local/singularity/3.0.0-dev/bin/singularity [tab] 
build       capability  exec        instance    keys        pull        push        run         shell       sign        verify      version
$ /usr/local/singularity/3.0.0-dev/bin/singularity -[tab]     
--debug       --quiet       --silent      --tokenfile=  --verbose     -d            -q            -s            -v
$ /usr/local/singularity/3.0.0-dev/bin/singularity ver[tab]
verify   version
$ /usr/local/singularity/3.0.0-dev/bin/singularity keys[tab] 
list     newpair  pull     push     search
```

yhcote commented on .gitignore:
```diff
@@ -9,6 +9,7 @@ singularity-*.tar.gz
 generate_docs
 
 .vscode
+etc/bash_completion.d/
```
Why does it have to be ignored?

jmstover replied:
If that got moved to $(BUILDDIR) then it can be removed. I believe it was just ignored because it was holding a generated file, so it wasn't something to be committed.

bauerm97 replied:
This folder should not be in `.gitignore`

yhcote commented on mlocal/frags/Makefile.stub:
```diff
@@ -73,6 +76,18 @@ $(go_OBJ): $(BUILDDIR)/vendors-done $(libruntime) $(BUILDDIR)/config.h
 	$(V)export BUILDDIR=`readlink -f $(BUILDDIR)` && cd $(SOURCEDIR)/src/pkg/buildcfg && go generate
 
+# bash_completion file
+$(bash_completion): $(go_OBJ)
```
Not sure that bash_completion target depends on $(go_OBJ)

ArangoGutierrez replied:
go run fails if you try to run bash_completion.go without the config.go file, so it needs to wait for it to be generated first

yhcote replied:
Seems like it's time to separate in a nicer way targets and their dependencies instead of having a unique set of go objs.

yhcote commented on mlocal/frags/Makefile.stub:
```diff
@@ -73,6 +76,18 @@ $(go_OBJ): $(BUILDDIR)/vendors-done $(libruntime) $(BUILDDIR)/config.h
+# bash_completion file
+$(bash_completion): $(go_OBJ)
+	@rm -f $(bash_completion)
+	@mkdir -p  $(BUILDDIR)/etc/bash_completion.d 
```
I think you can use $(@D) here instead if listing $(BUILDDIR)/etc/bash_completion.d

GodloveD commented on mlocal/frags/Makefile.stub:
```diff
@@ -31,18 +31,21 @@ go_BIN := $(singularity)
 go_INSTALL := $(singularity_INSTALL)
 
+bash_completion :=  $(BUILDDIR)/etc/bash_completion.d/singularity
+bash_completion_INSTALL := $(PREFIX)/etc/bash_completion.d/singularity
```
Need to remove `$(PREFIX)` here afaik so that the completion file just gets installed in  `/etc/bash_completion.d/singularity`.  Otherwise it won't be sourced when a new bash session is started.

jmstover replied:
*kicks @GodloveD *
https://github.com/singularityware/singularity/pull/1921#discussion_r215136112

yhcote replied:
I think we determined that installing everything in $(PREFIX) made for a cleaner and predictable install, where people interested in sourcing the completion file can easily throw that in their bash profile entries. I'd like to see some info around that though, but it seems like the better way to approach it. Rarely would software configured with --prefix should install stuff outside of that setup prefix.

GodloveD replied:
Blech.  If you insist I guess. What about making a `make completion` option or some such?

ArangoGutierrez replied:
As discussed, this will be handled via documentation.

GodloveD replied:
Boo!  😸

yhcote reviewed (changes requested):
Still a few thing here and there. Getting close!

bauerm97 commented on src/docs/bash_completion.d/bash_completion.go:
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) 2018, Sylabs Inc. All rights reserved.
```
This file should be move to `etc/bash_completion.d/bash_completion.go`

bauerm97 commented on src/docs/docs_gen.go:
```diff
@@ -0,0 +1,8 @@
+
+package docs
+
+//go:generate go run bash_completion.d/bash_completion.go "${BUILDDIR}/etc/bash_completion.d/singularity"
```
I don't think this file is strictly necessary. Would be much simpler to just directly call go run from the makefile itself, see the example for the `singularity.conf` file

bauerm97 commented on mlocal/frags/Makefile.stub:
```diff
@@ -73,6 +76,18 @@ $(go_OBJ): $(BUILDDIR)/vendors-done $(libruntime) $(BUILDDIR)/config.h
+$(bash_completion): $(go_OBJ)
+	@rm -f $(bash_completion)
+	@mkdir -p  $(BUILDDIR)/etc/bash_completion.d 
+	$(V)export BUILDDIR=`readlink -f $(BUILDDIR)` && cd $(SOURCEDIR)/src/docs && go generate
```
As I said, here we can just call `go run ...` directly, similarly to how we do for `singularity.conf` generation.

ArangoGutierrez committed:
remove etc/bash_completion from .gitignore

ArangoGutierrez committed:
use  instead of /etc/bash_completion.d

ArangoGutierrez committed:
use go run instead of go generate

ArangoGutierrez committed:
Merge remote-tracking branch 'upstream/master' into bash_completions

ArangoGutierrez committed:
mv bash_completion.go to etc/bash_completion.d

move bash_completion.go from src/docs/bash_completion.d to
etc/bash_completion.d

ArangoGutierrez committed:
mv bash_completion.go to etc/bash_completion.d

move bash_completion.go from src/docs/bash_completion.d to
etc/bash_completion.d

ArangoGutierrez committed:
Merge branch 'master' into bash_completions

ArangoGutierrez committed:
Merge branch 'bash_completions' of github.com:ArangoGutierrez/singularity into bash_completions

ArangoGutierrez committed:
better formating on line 49 Makefile.stub

bauerm97 reviewed (dismissed):
From my point of view the changes look good, but I would like to wait to merge for @yhcote to have a chance to give his thoughts as well

ArangoGutierrez committed:
Merge branch 'master' into bash_completions

yhcote commented on mlocal/frags/Makefile.stub:
```diff
@@ -42,11 +45,12 @@ cgo_CPPFLAGS = -I$(BUILDDIR) -I$(SOURCEDIR)/src/runtime -I$(SOURCEDIR)/src/runti
 
 INSTALLFILES := $(singularity_INSTALL) $(starter_INSTALL) $(starter_suid_INSTALL) $(sessiondir) \
-	$(config_INSTALL) $(dist_bin_SCRIPTS_INSTALL) $(capability_JSON) $(syecl_config_INSTALL)
+	$(config_INSTALL) $(dist_bin_SCRIPTS_INSTALL) $(capability_JSON) $(syecl_config_INSTALL)    \
```
why the extra spaces after $(syecl_config_INSTALL) and the \ ? There a few ago extra spaces down there too.

yhcote commented on mlocal/frags/Makefile.stub:
```diff
@@ -77,6 +81,18 @@ $(go_OBJ): $(BUILDDIR)/vendors-done $(libruntime) $(BUILDDIR)/config.h
 
+# bash_completion file
+$(bash_completion): $(go_OBJ)
+	@rm -f $(bash_completion)
```
Here use : @rm -f $@

yhcote commented on mlocal/frags/Makefile.stub:
```diff
@@ -77,6 +81,18 @@ $(go_OBJ): $(BUILDDIR)/vendors-done $(libruntime) $(BUILDDIR)/config.h
+$(bash_completion): $(go_OBJ)
+	@rm -f $(bash_completion)
+	@mkdir -p $(@D)
+	$(V) go run  $(SOURCEDIR)/etc/bash_completion.d/bash_completion.go $(bash_completion)
```
Remove extra space between $(V) and go
and
change $(bash_completion) at the end of the line for : $@

yhcote commented on mlocal/frags/Makefile.stub:
```diff
@@ -77,6 +81,18 @@ $(go_OBJ): $(BUILDDIR)/vendors-done $(libruntime) $(BUILDDIR)/config.h
+$(bash_completion_INSTALL): $(bash_completion)
+	@echo " INSTALL" $@
+	$(V)install -d $(@D)
+	$(V)install -m 0644 $(bash_completion) $(bash_completion_INSTALL)
```
Here change $(V)install -m 0644 $(bash_completion) $(bash_completion_INSTALL) for : 
$(V)install -m 0644 $< $@

yhcote reviewed (dismissed):
couple improvements.

ArangoGutierrez committed:
Merge branch 'master' into bash_completions

ArangoGutierrez committed:
Remove extra spaces on line 48 of Makefile.stub

ArangoGutierrez committed:
use @rm -f  on line 86 Makefile.stub

ArangoGutierrez committed:
better formating on make file bash_completion

Remove extra space between $(V) and go and
change $(bash_completion) at the end of the line for : $@

ArangoGutierrez committed:
Better formating of make file install bash_completion

change $(V)install -m 0644 $(bash_completion) $(bash_completion_INSTALL) for :
$(V)install -m 0644 $< $@

bauerm97 commented on mlocal/frags/Makefile.stub:
```diff
@@ -77,6 +81,18 @@ $(go_OBJ): $(BUILDDIR)/vendors-done $(libruntime) $(BUILDDIR)/config.h
 	$(V)export BUILDDIR=`readlink -f $(BUILDDIR)` && cd $(SOURCEDIR)/src/pkg/buildcfg && go generate
 
+# bash_completion file
+$(bash_completion): $(BUILDDIR)/config.h
```
This actually is incorrect, from what I understand this depends on the `buildcfg` package having already been generated,

ArangoGutierrez replied:
you are right, removed that commit

ArangoGutierrez committed:
Merge branch 'master' into bash_completions

GodloveD reviewed (changes requested):
Very easy change here.  I got tripped up when trying to test this.  Can you please remove https://github.com/singularityware/singularity/blob/master/etc/bash_completion.d/singularity

ArangoGutierrez committed:
remove etc/bash_completion.d/singularity

ArangoGutierrez committed:
Merge branch 'bash_completions' of github.com:ArangoGutierrez/singularity into bash_completions

GodloveD reviewed (approved):
lgtm.  I think that @yhcote is happy now too.
